### PR TITLE
fix(various): random album pagination fix - #749

### LINKED ISSUE #749: Random sort changes each time you switch pages
In random sort mode, clicking the next page button changes the random ordering, so some of the same albums show up. Maybe a random seed should be stored in the url?

### PR DESCRIPTION
**Pull request by AeronCyther**
# Closes
#749 

# Description
Added a way to seed the random sort used to order albums. 

# Changes
Added a middleware to handle random seeds in the form of cookies and add them to request's context when sorting albums by random. Added a context key to request. Implemented a pseudo-random function. Added a workaround for cases where random order mapping could not contain a comma (Example: "substr(id, 7)" turns into "order by substr(id asc, 7) asc" instead of "order by substr(id, 7) asc") by using a placeholder. 

# Notes
Currently the user stores the seed for 5 minutes since there is no way to differentiate between the api requests when the user has requested an album shuffle, and when the user just goes back to the first page (both requests only contain start and end). To get around this, the middleware looks for a "_shuffle" query. If found, it will generate a new seed. Once this is implemented, we may get rid of the 5 min expiry on the random seed.

### CODE REVIEW

**Comment by deluan** [comment]
>Added a workaround for cases where random order mapping could not contain a comma (Example: "substr(id, 7)" turns into "order by substr(id asc, 7) asc" instead of "order by substr(id, 7) asc") by using a placeholder.

Hey @AeronCyther, can you please describe a situation where this may happen? Which URL could cause this?

> Currently the user stores the seed for 5 minutes since there is no way to differentiate between the api requests when the user has requested an album shuffle, and when the user just goes back to the first page (both requests only contain start and end).

I don't think this is necessary: If the user goes back to the first page, it _is_ a new album shuffle

**Comment by AeronCyther** [comment]
> Hey @AeronCyther, can you please describe a situation where this may happen? Which URL could cause this?

@deluan Say for example, that my "random" function is ordering the entries by truncating their ids a bit. Then the random function could be "substr(id, 7)". But when squirrel gets that string as a parameter for SortBy, for some reason, it thinks that the comma is separating two sort parameters. An easy way to reproduce is to plug my random function in the master branch and debug print query after its creation in persistence/sql_base_repository.go. 

> I don't think this is necessary: If the user goes back to the first page, it _is_ a new album shuffle

Say I sort by random and I like an album on the first page, but I decide to go to next page to _just in case_ there's something better. I don't find anything I like so I go back to the first page using pagination (or even back button, but I don't see an easy solution for this), I wouldn't expect it to get randomised again.

**Comment by deluan** [comment]
> But when squirrel gets that string as a parameter for SortBy, for some reason, it thinks that the comma is separating two sort parameters.

I'll take a look at that, but maybe this is on our side? https://github.com/navidrome/navidrome/blob/6ee45a9ccc5e7ea4290c89030e67c99c0514bd25/persistence/sql_base_repository.go#L64-L96

**Comment by AeronCyther** [comment]
Seems like it is. I missed that. Either way, instead of trying to mess with that, since other code might depend on it, I decided to add a fallback instead.

**Comment by deluan** [comment]
Hey @AeronCyther, I just pushed a fix to the issue with functions in the sort order. Let me know if you want to finish this.

File: persistence/sql_base_repository.go [hunk 1]
```diff
@@ -144,6 +144,9 @@ func (r sqlRepository) queryAll(sq Sqlizer, response interface{}) error {
 	if err != nil {
 		return err
 	}
+	if _, ok := r.sortMappings["random-true"]; ok {
```
Comment (deluan) [inline persistence/sql_base_repository.go:147]: This should not be here, it is only related to albums, so should be in `album_repository`

File: server/server.go [hunk 2]
```diff
@@ -59,6 +59,7 @@ func (a *Server) initRoutes() {
 	r.Use(middleware.Heartbeat("/ping"))
 	r.Use(injectLogger)
 	r.Use(requestLogger)
+	r.Use(randomSeedMiddleware)
```
Comment (deluan) [inline server/server.go:62]: See comment in `middlewares.go`

**Review by deluan — CHANGES_REQUESTED** [review]

**Comment by AeronCyther** [comment]
@deluan I'll get to it ASAP

**Pull request closed without merging**